src/main/java/io/nats/client/impl/MessageQueue.java

```diff
@@ -27,39 +28,40 @@
     private final static int RUNNING = 1;
     private final static int DRAINING = 2;
 
-    public static final int MAX_SPINS = 200; // Default max spins
-    public static final int SPIN_WAIT = 50;
-    public static final int MAX_SPIN_TIME = SPIN_WAIT * MAX_SPINS;
-
     private final AtomicLong length;
     private final AtomicLong sizeInBytes;
     private final AtomicInteger running;
     private final boolean singleThreadedReader;
-    private final int maxSpins;
-    private final ConcurrentLinkedQueue<NatsMessage> queue;
-    private final ConcurrentLinkedQueue<Thread> waiters;
+    private final LinkedBlockingQueue<NatsMessage> queue;
     private final Lock filterLock;
 
-    MessageQueue(boolean singleReaderMode) {
-        this.queue = new ConcurrentLinkedQueue<>();
+    // Poison pill is a graphic, but common term for an item that breaks loops or forces some other action
+    // In this class the poisonPill is used to break out of timed waits on the blocking queue.
+    private final NatsMessage poisonPill;
+
+    /**
+     * If publishHighwaterMark is set to 0 the underlying queue can grow forever. This is used by readers
+     * to prevent the read thread from blocking. If set to a number, the publish command will block, which provides
+     * backpressure on a publisher if the writer is slow to push things onto the network.
+     * @param singleReaderMode allows the use of "accumulate"
+     * @param publishHighwaterMark sets a limit on the size of the underlying queue
+     */
+    MessageQueue(boolean singleReaderMode, int publishHighwaterMark) {
+        this.queue = publishHighwaterMark > 0 ? new LinkedBlockingQueue<NatsMessage>(publishHighwaterMark) : new LinkedBlockingQueue<NatsMessage>();

```

Comment:
Well, actually since it is a linked blocking queue, there is always a capacity, which in this case looking at javadoc, if no capacity is specified it is 2^31-1 :-). Nothing to change here, just mentioning it.